### PR TITLE
Add Tool-Centric APIs to the Tool Shed 2.0

### PR DESCRIPTION
**Pull request by jmchilton**
Adds the following backend enhancements to ``galaxy-tool-util``...

- Brings in model layer from structured tool state work to describe tool inputs. 
- Reworks citation handling for proper separation of parsing into ``galaxy.tool_util.parser``.
- Implement a variety of Pydantic models to describe tool outputs.

... and brings those together to add two new APIs to the Tool Shed 2.0 designed for reasoning about tools outside of Galaxy.

- ``/api/tools/{trs_tool_id}/version/{version}`` returns a parsed version of various information from the ``ToolSource`` abstraction that is de-coupled from XML and not dependent on Galaxy's ``Tool`` classes. The model for this information is:

```python
class ParsedTool(BaseModel):
    id: str
    version: Optional[str]
    name: str
    description: Optional[str]
    inputs: List[ToolParameterT]
    outputs: List[ToolOutput]
    citations: List[Citation]
    license: Optional[str]
    profile: Optional[str]
    edam_operations: List[str]
    edam_topics: List[str]
    xrefs: List[XrefDict]
    help: Optional[str]
```
- ``/api/tools/{trs_tool_id}/version/{version}/parameter_request_schema`` returns a JSON schema that can be used to validate a tool inputs according to the API that will be added with the Galaxy internals portion of the structured tool state PR #17393. This is generated from the metadata model that is exposed via the other API as the ``inputs`` field.

This is better than the last attempt #18470 for a variety of reasons - the tool state work has advanced and ``repeat`` min and max instance counts are now represented by the schema as are ``section`` elements, the tool shed API is much richer and more harmonious (modeling inputs and outputs together should allow for full workflow validation by external tooling, including citations, xrefs, and help in the API should enhance our ability to provide information about tools from sources such the IDE plugin), the caching has been reworked to be more future facing and gracefully handle model changes, and finally this version exposes stock tools as well.

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

### CODE REVIEW

File: doc/source/dev/tool_state.md
Comment (mvdbeek) [inline doc/source/dev/tool_state.md:1]: Did you mean to put content here ?

File: lib/galaxy/tool_util/parameters/_types.py
Comment (mvdbeek) [inline lib/galaxy/tool_util/parameters/_types.py]:
```suggestion
# https://stackoverflow.com/questions/56832881/check-if-a-field-is-typing-optional
from typing_extensions import (
    get_args,
    get_origin,
)
```
?

File: lib/galaxy/tool_util/parameters/case.py
Comment (mvdbeek) [inline lib/galaxy/tool_util/parameters/case.py]: Unqualified input references are a big thing unfortunately. Is that going to break on each of these arguably broken test cases ? And there might also be nested inputs (for sections, conditionals etc), how are these handled ?

File: lib/galaxy/tool_util/parameters/case.py
Comment (mvdbeek) [inline lib/galaxy/tool_util/parameters/case.py]: That probably also needs to be run at each level of nesting ? I can verify how that works.

File: lib/galaxy/tool_util/parameters/factory.py
Comment (mvdbeek) [inline lib/galaxy/tool_util/parameters/factory.py:51]: Is that the right layer to provide default values ? Maybe produce optional string instead ?

File: lib/galaxy/tool_util/parameters/factory.py
Comment (mvdbeek) [inline lib/galaxy/tool_util/parameters/factory.py:71]: This makes me wonder if we should consume parameter models in basic.py to avoid any drift.

File: lib/galaxy/tool_util/parameters/factory.py
Comment (mvdbeek) [inline lib/galaxy/tool_util/parameters/factory.py]: Should go to top eventually

File: lib/galaxy/tool_util/parameters/factory.py
Comment (mvdbeek) [inline lib/galaxy/tool_util/parameters/factory.py]:
```suggestion
```
seems done ?

File: lib/galaxy/tool_util/parameters/models.py
Comment (mvdbeek) [inline lib/galaxy/tool_util/parameters/models.py]: This serves no function anymore (AFAICT). I guess it's a good test case for how we evolved models ?

**Review by mvdbeek — COMMENTED** [review]
Looks great, just some minor comments in line. Can't wait to play with this.

File: lib/galaxy/tool_util/parameters/_types.py
Comment (jmchilton) [reply lib/galaxy/tool_util/parameters/_types.py]: tool_util should support Python 3.7 IMO

File: lib/galaxy/tool_util/parameters/case.py
Comment (jmchilton) [reply lib/galaxy/tool_util/parameters/case.py]: That whole file is WIP. I wouldn't worry about it - it uses stuff from the full structured tool state branch. I was never able to get the test case model working in that branch but it was my next up task. I can pull this out and put it in the other branch.

File: lib/galaxy/tool_util/parameters/_types.py
Comment (mvdbeek) [reply lib/galaxy/tool_util/parameters/_types.py]: typing_extensions is a backport of newer type system features to older python versions. This seems to work fine on 3.7:
```
❯ docker run -it python:3.7 sh
# pip install typing_extensions
Collecting typing_extensions
  Downloading typing_extensions-4.7.1-py3-none-any.whl (33 kB)
Installing collected packages: typing_extensions
Successfully installed typing_extensions-4.7.1
WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv

[notice] A new release of pip is available: 23.0.1 -> 24.0
[notice] To update, run: pip install --upgrade pip
# python -c "from typing_extensions import get_args, get_origin"
# echo $?
0
```

File: lib/galaxy/tool_util/parameters/factory.py
Comment (jmchilton) [reply lib/galaxy/tool_util/parameters/factory.py:71]: We still haven't even decoupled basic.py from XML fully yet 😅 . I think in the meantime the tool test suite is what keeps things on track. This is why implementing the test cases was such a goal for me. Hmm... I am not sure I think we should block on this but I think it is a worthy goal. I could spend a day looking at this before merging this if you'd like... I must have had thoughts in this direction based on implementing that ``get_color_value`` function in here.

File: lib/galaxy/tool_util/parameters/_types.py
Comment (jmchilton) [reply lib/galaxy/tool_util/parameters/_types.py]: I thought you were suggesting dropping Python 3.7 not using typing_extensions. Sorry - I will try this.

File: lib/galaxy/tool_util/parameters/factory.py
Comment (mvdbeek) [reply lib/galaxy/tool_util/parameters/factory.py:71]: Happy to keep this as is for now. That's a side branch I can pull out too after merging the initial PR.

File: doc/source/dev/tool_state.md
Comment (jmchilton) [reply doc/source/dev/tool_state.md:1]: I was going to stick the two images in there and do a short writeup but I never got around to that. I'll remove this file. The images are a little questionable without linking them but I think they capture some good documentation.

File: lib/galaxy/tool_util/parameters/factory.py
Comment (jmchilton) [reply lib/galaxy/tool_util/parameters/factory.py:71]: I'll create an issue to track this discussion.

File: lib/galaxy/tool_util/parameters/case.py
Comment (jmchilton) [reply lib/galaxy/tool_util/parameters/case.py]: This file has been moved to https://github.com/galaxyproject/galaxy/pull/17393. I'm going to resolve these two comments.

File: lib/galaxy/tool_util/parameters/factory.py
Comment (jmchilton) [reply lib/galaxy/tool_util/parameters/factory.py:71]: https://github.com/galaxyproject/galaxy/issues/18537

**Comment by hexylena** [comment]
(Just want to say this is fantastic to see, it'll let me throw away a repository where I collect this data from live servers as the old TS doesn't expose it and I need this metadata. happy day. Thank you @jmchilton!!)

**Review by mvdbeek — APPROVED** [review]
Looks great, let's deploy it. Can you regenerate the conflicting schema ?

**Comment by mvdbeek** [comment]
This is so awesome, thanks a lot @jmchilton! This will enable so many cool things for a long time to come!

**Pull request merged**
